lib/Refactoring/AddExplicitCodableImplementation.cpp

```diff
@@ -12,183 +12,230 @@
-    Options.ExcludeAttrList.push_back(DeclAttrKind::HasInitialValue);
-
-    Printer.printNewline();
-    D->print(Printer, Options);
-
-    return Action::SkipNode();
+    // Look for synthesized witness decls and print them.
+    for (auto *conformance : IDC->getLocalConformances()) {
+      auto protocol = conformance->getProtocol();
+      auto kind = protocol->getKnownProtocolKind();
+      if (kind == KnownProtocolKind::Encodable ||
+          kind == KnownProtocolKind::Decodable) {
+        for (auto requirement : protocol->getProtocolRequirements()) {
+          auto witness = conformance->getWitnessDecl(requirement);
+          if (witness->isSynthesized()) {
+            printMember(witness);
+            Printer.printNewline();
+          }
+        }
+      }
+    }
   }
-};
 
-void AddCodableContext::printInsertionText(ResolvedCursorInfoPtr CursorInfo,
-                                           SourceManager &SM,
-                                           llvm::raw_ostream &OS) {
-  StringRef ExtraIndent;
-  StringRef CurrentIndent =
-      Lexer::getIndentationForLine(SM, getInsertStartLoc(), &ExtraIndent);
-  std::string Indent;
-  if (getInsertStartLoc() == StartLoc) {
-    Indent = (CurrentIndent + ExtraIndent).str();
-  } else {
-    Indent = CurrentIndent.str();
+public:
+  static AddCodableContext getFromCursorInfo(ResolvedCursorInfoPtr Info);
+
+  bool isApplicable() const {
+    if (!IDC)
+      return false;
+
+    // Check if 'IDC' conforms to 'Encodable' and/or 'Decodable' and any of the
+    // requirements are synthesized.
+    for (auto *conformance : IDC->getLocalConformances()) {
+      auto protocol = conformance->getProtocol();
+      auto kind = protocol->getKnownProtocolKind();
+      if (kind == KnownProtocolKind::Encodable ||
+          kind == KnownProtocolKind::Decodable) {
+        // Check if any of the protocol requirements are synthesized.
+        for (auto requirement : protocol->getProtocolRequirements()) {
+          auto witness = conformance->getWitnessDecl(requirement);
+          if (!witness || witness->isSynthesized())
+            return true;
+        }
+      }
+    }
+    return false;
   }
 
-  ExtraIndentStreamPrinter Printer(OS, Indent);
-  Printer.printNewline();
-  SynthesizedCodablePrinter Walker(Printer);
-  DC->getAsDecl()->walk(Walker);
-}
+  void getInsertion(SourceLoc &insertLoc, std::string &insertText) const {
+    insertLoc = getInsertStartLoc();
+    llvm::raw_string_ostream OS(insertText);
+    printInsertText(OS);
+  }
+};
 
 AddCodableContext
-AddCodableContext::getDeclarationContextFromInfo(ResolvedCursorInfoPtr Info) {
+AddCodableContext::getFromCursorInfo(ResolvedCursorInfoPtr Info) {
   auto ValueRefInfo = dyn_cast<ResolvedValueRefCursorInfo>(Info);
   if (!ValueRefInfo) {
-    return AddCodableContext();
+    return nullptr;
   }
+
+  if (auto *ext = ValueRefInfo->getExtTyRef()) {
+    // For 'extension Outer.Inner: Codable {}', only 'Inner' part is valid.
```

Comment:
Ideally (IMO) anything around the extension (or decl) would be valid, ie. the keyword, any part of the name, the colon... But I suspect that would be fairly different to others we have today. Plus this is better than what we have now so 🤷